### PR TITLE
Add initial database migration manager

### LINKED ISSUE #93: Database migrations are manual and difficult to version
We should include a versioned migration system for our primary datastore `postgres` service. However, since the service can be used by multiple other services, the migration handler should reside outside of any language-specific ORM, and use plain SQL preferably.

### PR DESCRIPTION
**Pull request by jazeved0**
Co-authored-by: @AlexAppleManzer 

## Description

Attempts to solve the issues described in #93 involved with running database migrations that have to be performed manually by our DBA on the production server. This PR modifies the entrypoint of the `postgres` service's container to run migrations as necessary at runtime. This is similar to the native behavior of the postgres image to [run initialization scripts](https://github.com/docker-library/docs/blob/master/postgres/README.md#initialization-scripts). However, these mechanisms are insufficient for migrations as they only run on the first starting of the database server, and never again. Instead, we need a system that automatically versions and tracks which migrations have been applied at each run, running all new migrations as neccessary. Note that it is necessary to run these migrations at *runtime* instead of *built time* due to the fact that they need to persist to the server's permanent volume that is only available at runtime (and they shouldn't destroy old data).

### Implementation details

The pull request works by using a special Python script, `bootstrap.py`, that modifies the contents of the [standard entrypoint shell script provided by the Postgres image](https://github.com/docker-library/postgres/blob/682ff83c5c83f1b6f2b02caf7aa3e17a491b403a/13/docker-entrypoint.sh), adding an unconditional call to the same script with a special mode flag directly before the start of the server for general availability. This means the migrations will *always* be checked and possibly run, and this also occurs after the initial database setup (in case the database had no content beforehand). When the script detects the special mode flag, it instead runs the checking process, which consists of the following steps:

1. Gets all SQL files in the `migrations` folder
2. Detects if there is a file in the `/db/current_migration` folder (mounted as a persistent volume for the container), reading its contents if it exists
3. Gets a list of all migration files that are more recent than the last-run migration from the above file (or all migrations if no migrations have previously been run)

If any migrations are found, then the script:

1. Invokes a subroutine in the original entrypoint shell script that starts a temporary postgres server using `pg_ctl` that isn't available to the network
2. Runs each SQL file on the server, performing the migrations and echoing the output to stdout (using a pre-existing subroutine)
3. Shuts down the temporary server using another pre-existing subroutine

Finally, the script exits the special mode-flag version of the Python script, yielding control back to the original entrypoint shell script, which starts the general availability server process immediately afterwards.

This means that the flow of the active process looks like the following (assuming migrations are being run):

```
Docker container entrypoint
└─┬─ `python3 /db/bootstrap.py`
  └─┬─ `bash /db/entrypoint.sh`
    ├─── (...)
    └─┬─ `python3 /db/bootstrap.py --migrate`
      └┬── `pg_ctl start`
       ├── `psql`
      ┌┴── `pg_ctl stop`
    ┌─┴─ `python3 /db/bootstrap.py --migrate`
    ├─ `bash /db/entrypoint.sh`
    └─ `postgres` (until container exit)
```

## Checklist

- [x] [Linked PR to corresponding issue in this repo](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue)
- [x] Manually tested locally
- [x] Additional issue created in [docs repo](https://github.com/architus/docs.archit.us/issues) (if necessary): https://github.com/architus/docs.archit.us/issues/13

### CODE REVIEW

File: db/bootstrap.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+ORIGINAL_ENTRYPOINT = "/usr/local/bin/docker-entrypoint.sh"
+INJECTED_ENTRYPOINT = os.path.join(BASE_DIR, "./entrypoint.sh")
+
+
+def main():
+    print("Running db migration manager")
+    current_migration = get_current_migration()
+    to_run = get_new_migrations(current_migration)
+    if to_run:
+        print(f"Running migrations: [{', '.join(to_run)}]")
```
Comment (johnyburd) [inline db/bootstrap.py:29]: isn't this the exact same behavior as `f"Running migrations: {to_run}"`

File: db/Dockerfile [hunk 2]
```diff
@@ -1,7 +1,18 @@
+# Debian variant
 FROM postgres
```
Comment (johnyburd) [inline db/Dockerfile:2]: we should probably make this a specific version since we're messing with internal stuff

**Review by johnyburd — CHANGES_REQUESTED** [review]

File: db/bootstrap.py (hunk 1 above)
Comment (AlexAppleManzer) [reply db/bootstrap.py:29]: probably

File: db/Dockerfile (hunk 2 above)
Comment (AlexAppleManzer) [reply db/Dockerfile:2]: ok

**Review by AlexAppleManzer — CHANGES_REQUESTED** [review]
does not run migrations properly on subsquent builds

**Comment by johnyburd** [comment]
is this ded?

**Comment by jazevedo620** [comment]
Moved to #97

**Pull request closed without merging**